# **Analyse Exploratoire des Données (EDA)**
## **Cas d’étude – Stage Data Engineering chez Artefact CI**

### Nom & Prénoms : **BALOGUN John Oluwasegun**  
**Objectif :** Analyse exploratoire et compréhension métier du jeu de données  
**Contexte :** Préparation à la modélisation relationnelle (3FN)

## Initialisation 

In [30]:
import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings("ignore") # Masque les warnings pour une lecture plus fluide

pd.set_option("display.max_columns", None) # Affiche toutes les colonnes des DataFrames


## 1. EXAMEN DE LA STRUCTURE DU FICHIER

### 1.1 Chargement du dataset

In [31]:
file_path = r"C:\Users\VICTUS\Downloads\ARTEFACT\fashion_store_sales.csv"
df = pd.read_csv(file_path)

print("Dimensions du dataset")
print(f"Lignes   : {df.shape[0]:,}")
print(f"Colonnes : {df.shape[1]}")

Dimensions du dataset
Lignes   : 2,253
Colonnes : 29


In [32]:
df.head()

sale_date  item_id  sale_id  product_id  quantity  original_price  \
0  2025-06-16     2270      658         403         1           81.80   
1  2025-06-17     1170      336         284         1           81.79   
2  2025-04-16     2496     1255          71         1           80.76   
3  2025-05-06     1273      331          98         1           78.52   
4  2025-06-15     1829     1079          98         1           78.52   

   unit_price  discount_applied discount_percent  discounted  item_total  \
0       81.80              0.00            0.00%           0       81.80   
1       81.79              0.00            0.00%           0       81.79   
2       80.76              0.00            0.00%           0       80.76   
3       78.52              0.00            0.00%           0       78.52   
4       78.52              0.00            0.00%           0       78.52   

      channel channel_campaigns  total_amount            product_name  \
0  App Mobile        App Mobile        374.25     Elegant Satin Dress   
1  E-commerce    Website Banner        536.47  Essential Cotton Shoes   
2  App Mobile        App Mobile        104.81  Modern Ribbed Trousers   
3  App Mobile        App Mobile        263.87       Modern Boxy Shoes   
4  App Mobile        App Mobile        173.84       Modern Boxy Shoes   

  category brand  color size  catalog_price  cost_price  customer_id  gender  \
0  Dresses  Tiva    Red    L          81.80       45.12          835  Female   
1    Shoes  Tiva  White   35          81.79       35.02          790  Female   
2    Pants  Tiva    Red   XL          80.76       51.01          464  Female   
3    Shoes  Tiva  Black   38          78.52       41.48          100  Female   
4    Shoes  Tiva  Black   38          78.52       41.48          837  Female   

  age_range signup_date first_name  last_name                    email  \
0     46-55  2025-04-26      Dusty  Comerford   dcomerfordn6@google.nl   
1     16-25  2025-04-26      Beale      Seeds       bseedslx@phpbb.com   
2     36-45  2025-04-14       Juan  Blacklock  jblacklockcv@discuz.net   
3     26-35  2025-01-30     Godfry  Cockerill   gcockerill2r@vimeo.com   
4     46-55  2025-03-02        NaN      Kilby   lkilbyn8@wordpress.com   

    country  
0  Portugal  
1    France  
2   Germany  
3     Italy  
4   Germany

### 1.2 Liste et typologie des colonnes

In [33]:
colonnes = pd.DataFrame({
    "colonne": df.columns,
    "type": df.dtypes.values
})

colonnes

colonne     type
0           sale_date   object
1             item_id    int64
2             sale_id    int64
3          product_id    int64
4            quantity    int64
5      original_price  float64
6          unit_price  float64
7    discount_applied  float64
8    discount_percent   object
9          discounted    int64
10         item_total  float64
11            channel   object
12  channel_campaigns   object
13       total_amount  float64
14       product_name   object
15           category   object
16              brand   object
17              color   object
18               size   object
19      catalog_price  float64
20         cost_price  float64
21        customer_id    int64
22             gender   object
23          age_range   object
24        signup_date   object
25         first_name   object
26          last_name   object
27              email   object
28            country   object

### 1.3 Analyse des types et conversions attendues

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2253 entries, 0 to 2252
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sale_date          2253 non-null   object 
 1   item_id            2253 non-null   int64  
 2   sale_id            2253 non-null   int64  
 3   product_id         2253 non-null   int64  
 4   quantity           2253 non-null   int64  
 5   original_price     2253 non-null   float64
 6   unit_price         2253 non-null   float64
 7   discount_applied   2253 non-null   float64
 8   discount_percent   2253 non-null   object 
 9   discounted         2253 non-null   int64  
 10  item_total         2253 non-null   float64
 11  channel            2253 non-null   object 
 12  channel_campaigns  2253 non-null   object 
 13  total_amount       2028 non-null   float64
 14  product_name       2253 non-null   object 
 15  category           2253 non-null   object 
 16  brand              2253 

### 1.4 Cardinalité par colonne (fondamental pour la modélisation)

In [35]:
cardinalite = pd.DataFrame({
    "colonne": df.columns,
    "valeurs_uniques": df.nunique(),
    "pourcentage": (df.nunique() / len(df)) * 100
}).sort_values("valeurs_uniques", ascending=False)

cardinalite

colonne  valeurs_uniques  pourcentage
item_id                      item_id             2253       100.00
item_total                item_total             1555        69.02
sale_id                      sale_id              905        40.17
total_amount            total_amount              896        39.77
unit_price                unit_price              638        28.32
customer_id              customer_id              580        25.74
last_name                  last_name              564        25.03
first_name                first_name              535        23.75
email                          email              523        23.21
product_name            product_name              499        22.15
product_id                product_id              499        22.15
original_price        original_price              475        21.08
catalog_price          catalog_price              475        21.08
cost_price                cost_price              455        20.20
discount_applied    discount_applied              182         8.08
signup_date              signup_date              140         6.21
sale_date                  sale_date               51         2.26
size                            size                9         0.40
country                      country                6         0.27
quantity                    quantity                5         0.22
category                    category                5         0.22
age_range                  age_range                5         0.22
color                          color                5         0.22
channel_campaigns  channel_campaigns                4         0.18
discount_percent    discount_percent                3         0.13
channel                      channel                2         0.09
discounted                discounted                2         0.09
brand                          brand                1         0.04
gender                        gender                1         0.04

### 1.5 Interprétation métier de la cardinalité

In [36]:
haute_card = cardinalite[cardinalite["pourcentage"] > 90]
basse_card = cardinalite[cardinalite["pourcentage"] < 10]

print("Colonnes à très forte cardinalité")
haute_card

Colonnes à très forte cardinalité


colonne  valeurs_uniques  pourcentage
item_id  item_id             2253       100.00

#### Interprétation de la colonne `item_id`

La colonne `item_id` présente une cardinalité de 100 %, ce qui indique
qu’elle est unique pour chaque ligne du dataset.

Cette colonne correspond à un identifiant technique de ligne et ne représente aucune entité métier.

`item_id` ne doit pas être utilisée comme clé primaire fonctionnelle
dans le modèle de données cible.

## 2. IDENTIFICATION DES ENTITÉS MÉTIER

### 2.1 Entité VENTES (table de faits)

In [37]:
df["sale_id"].nunique(), len(df)

(905, 2253)

#### Granularité des ventes

L’analyse montre que le nombre de `sale_id` uniques (905) est inférieur
au nombre total de lignes (2 253).

Cela indique qu’une vente peut contenir plusieurs lignes,
correspondant à plusieurs produits dans un même panier.

La granularité du dataset est donc **la ligne de vente**.

Ce point est structurant pour la modélisation relationnelle :
- `sale_id` ne peut pas être une clé primaire seule
- une clé composite ou une clé technique sera nécessaire


### 2.2 Entité PRODUITS

In [38]:
df["product_id"].nunique()

499

#### Vérification des dépendances fonctionnelles

In [39]:
df.groupby("product_id").agg({
    "product_name": "nunique",
    "category": "nunique",
    "brand": "nunique",
    "color": "nunique",
    "size": "nunique"
}).query("product_name > 1 or category > 1")

Empty DataFrame
Columns: [product_name, category, brand, color, size]
Index: []

Résultat vide, prouve que : product_id → product_name, category, brand, color, size donc Produit est une entité stable, parfait pour une table PRODUITS en 3FN

### 2.3 Entité CLIENTS

In [40]:
df["customer_id"].nunique()

580

In [41]:
df.groupby("customer_id").agg({
    "email": "nunique",
    "country": "nunique",
    "age_range": "nunique"
}).query("email > 1 or country > 1")


Empty DataFrame
Columns: [email, country, age_range]
Index: []

Résultat vide également, conclusion : customer_id → email, country, age_range donc aucun conflit métier et une Table CLIENTS propre

### 2.4 Entités CANAUX & CAMPAGNES

In [42]:
df[["channel", "channel_campaigns"]].drop_duplicates().sort_values("channel")

channel channel_campaigns
0    App Mobile        App Mobile
163  App Mobile      Social Media
1    E-commerce    Website Banner
233  E-commerce             Email

## 3. REDONDANCES & ANOMALIES

### 3.1 Valeurs manquantes

In [43]:
missing = pd.DataFrame({
    "colonne": df.columns,
    "nb_manquants": df.isnull().sum(),
    "pourcentage": (df.isnull().sum() / len(df)) * 100
}).query("nb_manquants > 0").sort_values("nb_manquants", ascending=False)

missing

colonne  nb_manquants  pourcentage
total_amount  total_amount           225         9.99
email                email           224         9.94
first_name      first_name           116         5.15
last_name        last_name            64         2.84

### 3.2 Colonnes constantes (non informatives)

In [44]:
[col for col in df.columns if df[col].nunique() == 1]

['brand', 'gender']

Ces colonnes ne portent aucune information discriminante et peuvent être supprimées

### 3.3 Redondance des prix

In [49]:
print(f"{(df['original_price'] == df['catalog_price']).mean() * 100:.0f} %")

100 %


Une seule colonne à conserver

### 3.4 Cohérence des montants

In [46]:
df["item_total_calc"] = df["quantity"] * df["unit_price"]

df[np.abs(df["item_total"] - df["item_total_calc"]) > 0.01].shape[0]


77

77 lignes présentent une incohérence entre le champ item_total et le
montant recalculé (quantity × unit_price).

Le champ item_total ne peut donc pas être considéré comme fiable et
devra être recalculé lors de l’ingestion des données.

### 3.5 Cohérence des remises

In [47]:
df["discount_percent_num"] = (
    df["discount_percent"]
    .str.replace("%", "", regex=False)
    .astype(float)
)

In [48]:
df[
    ((df["discount_percent_num"] > 0) & (df["discounted"] == 0)) |
    ((df["discount_percent_num"] == 0) & (df["discounted"] == 1))
]

Empty DataFrame
Columns: [sale_date, item_id, sale_id, product_id, quantity, original_price, unit_price, discount_applied, discount_percent, discounted, item_total, channel, channel_campaigns, total_amount, product_name, category, brand, color, size, catalog_price, cost_price, customer_id, gender, age_range, signup_date, first_name, last_name, email, country, item_total_calc, discount_percent_num]
Index: []

discounted et discount_percent sont cohérents donc la logique métier est respectée

### CONCLUSION – ANALYSE EXPLORATOIRE

Le jeu de données correspond à un dataset e-commerce transactionnel
fortement dénormalisé, avec une granularité au niveau ligne de vente.

L’analyse met en évidence :
- des ventes multi-lignes (panier),
- des dépendances fonctionnelles claires pour les entités Clients et Produits,
- des redondances importantes (informations clients, produits, prix),
- des colonnes non informatives (brand, gender),
- des anomalies de calcul sur le champ item_total.

Ces constats justifient pleinement une normalisation du modèle de données
jusqu’à la troisième forme normale (3FN), qui sera abordée dans la section suivante.
